lms/djangoapps/shoppingcart/tests/test_views.py

```diff
@@ -1272,6 +1252,42 @@ def test_disabled_paid_courses(self):
         self._assert_404(reverse('shoppingcart.views.reset_code_redemption', args=[]), use_post=True)
         self._assert_404(reverse('shoppingcart.views.billing_details', args=[]))
 
+    def test_upgrade_postpay_callback_emits_ga_event(self):
+        # Create other carts first
+
+        Order.get_cart_for_user(self.user).start_purchase()
+        # Purchase a verified certificate
+        self.cart = Order.get_cart_for_user(self.user)
+        CertificateItem.add_to_order(self.cart, self.verified_course_key, self.cost, 'honor')
+        self.cart.start_purchase()
+
+        self.login_user()
+        # setting the attempting upgrade session value.
+        s = self.client.session
+        s['attempting_upgrade'] = True
+        s.save()
+
+        # Simulate hitting the post-pay callback
+        with patch('shoppingcart.views.process_postpay_callback') as mock_process:
+            mock_process.return_value = {'success': True, 'order': self.cart}
+            url = reverse('shoppingcart.views.postpay_callback')
+
+            self.assertTrue(self.client.session.get('attempting_upgrade'))
+            __ = self.client.post(url, follow=True)
+            self.assertFalse(self.client.session.get('attempting_upgrade'))
+
+            course_enrollment = CourseEnrollment.get_or_create_enrollment(self.user, self.course_key)
+            course_enrollment.emit_event('edx.course.enrollment.upgrade.succeeded')
```

Comment:
Why are these lines necessary?  Shouldn't the event be fired by the POST request?